app/scripts/modules/core/src/presentation/hooks/useDataSource.hook.ts

```diff
@@ -0,0 +1,32 @@
+import { useObservableValue } from './useObservableValue.hook';
+import { useLatestCallback } from './useLatestCallback.hook';
+
+import { ApplicationDataSource } from '../../application';
+
+export interface IDataSourceResult<T> {
+  data: T;
+  refresh: () => void;
+}
+
+/**
+ * A react hook that returns the current value an ApplicationDataSource
+ * and triggers a re-render when a new value is set. Also provides a function
+ * for refreshing the data source manually.
+ *
+ * @param dataSource the data source to subscribe to
+ * @returns IDataSourceResult<T>
+ */
+export const useDataSource = <T>(dataSource: ApplicationDataSource<T>): IDataSourceResult<T> => {
+  const data = useObservableValue(dataSource.data$, dataSource.data);
+
+  // Memoize to give consumers a stable ref,
+  // but don't return the promise that dataSource.refresh() returns
+  const refresh = useLatestCallback(() => {
+    dataSource.refresh();
+  });
+
+  return {
```

Comment:
Should we make this a tuple?

Reply:
Just kinda like
```
const [environments, refreshEnvironments] = useDataSource(dataSource);
```
more than
```
const { data: environments, refresh: refreshEnvironments } = useDataSource(dataSource);
```

Reply:
Fine either way though

Reply:
i also have a mild pref for tuples in these APIs, but in this case i wanted to leave room for adding additional nice things onto this API without a big refactor (e.g. what if you want to know what the pending/in flight status of a refresh is? errors?)

Reply:
A 3-tuple then?

Just kidding, let's just leave it as is then.

Reply:
Yeah, FWIW I went through this same process on `useLatestPromise`.  Started out with tuples.  Then threeples.  Then returned values that some callers wouldn't care about, (but they might care about a latter fourple and have to destructure the thirdple to a throwaway variable).